[CORE-14557] `storage`: increment `_stats.fsyncs` in `segment_appender::flush()`

WillemKauf opened the pull request:
We increment `_stats.fsyncs` elsewhere we call `_out.flush()`, but not at this call site.

Amend the call site to match the others.

This seems to have manifested in a test failure:

```
src/v/storage/tests/segment_appender_test.cc:237: Failure
Expected: (appender->get_stats().fsyncs) >= (1), actual: 0 vs 1
```

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

WillemKauf commented on src/v/storage/segment_appender.cc:
```diff
@@ -654,6 +654,7 @@ ss::future<> segment_appender::flush() {
       *this);
 
     return _out.flush().handle_exception([this](std::exception_ptr e) {
+        ++_stats.fsyncs;
```
Good bot

WillemKauf commented:
Force push to:
* Fix exception handling (`throw` won't be caught by a `f().then().handle_exception()` block)

WillemKauf commented:
```
bazel test //src/v/storage/tests:segment_appender_test --cache_test_results=no --runs_per_test=1000
src/v/storage/tests/segment_appender_test.cc:238: Failure
Expected: (appender->get_stats().fsyncs) <= (2), actual: 65 vs 2
``` 

:thinking:

WillemKauf commented:
Removing check for `_stats.fsyncs <= 2`. This flakes at a rate of about 1/1000 on my machine.

@mmaslankaprv I'm not sure if this invalidates any expectations you had for your changes in https://github.com/redpanda-data/redpanda/pull/28230.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--3b5eb417-056a-4bea-978e-2281676ea0d5-->
<!--TEST RESULTS-->
<!--46f5b77c-f223-4c47-ae8a-15afef225e88-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75782">build#75782</a></summary>


| test_class                           | test_method                | test_arguments                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                  |
|:-------------------------------------|:---------------------------|:-----------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkBasicTests                 | test_task_pausing          | {"shuffle_leadership": true}                                                 | integration | https://buildkite.com/redpanda/redpanda/builds/75782#019a5b06-4cd5-499c-a652-8516992c1a94 | FLAKY         | 19/21    | upstream reliability is '98.24561403508771'. current run reliability is '90.47619047619048'. drift is 7.76942 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_task_pausing                          |
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/75782#019a5b07-cad9-41e7-8792-b8fe43747369 | FLAKY         | 20/21    | upstream reliability is '98.9672977624785'. current run reliability is '95.23809523809523'. drift is 3.7292 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync |
| ShardPlacementTest                   | test_core_count_change     | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/75782#019a5b06-4cd5-499c-a652-8516992c1a94 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShardPlacementTest&test_method=test_core_count_change                       |
| WriteCachingFailureInjectionE2ETest  | test_crash_all             | {"use_transactions": false}                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/75782#019a5b06-4cd3-4237-81e9-afd93b48336e | FLAKY         | 17/21    | upstream reliability is '90.5587668593449'. current run reliability is '80.95238095238095'. drift is 9.60639 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all              |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75946">build#75946</a></summary>


| test_class                                        | test_method                                       | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                               |
|:--------------------------------------------------|:--------------------------------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests                     | test_replication_basic                            | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}} | integration | https://buildkite.com/redpanda/redpanda/builds/75946#019a6fca-66c3-4f29-84c5-a8a78a3f98b2 | FLAKY         | 16/21    | upstream reliability is '91.12187247780467'. current run reliability is '76.19047619047619'. drift is 14.9314 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic         |
| ShadowLinkingReplicationTests                     | test_replication_with_failures                    | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/75946#019a6f71-0ce4-4e85-bb4b-5a2055d405c6 | FLAKY         | 17/21    | upstream reliability is '96.74952198852772'. current run reliability is '80.95238095238095'. drift is 15.79714 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures |
| MountUnmountIcebergTest                           | test_simple_remount                               | {"cloud_storage_type": 1}                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/75946#019a6f71-0ce4-4e85-bb4b-5a2055d405c6 | FLAKY         | 16/21    | upstream reliability is '90.0'. current run reliability is '76.19047619047619'. drift is 13.80952 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount                  |
| NodesDecommissioningTest                          | test_decommissioning_working_node                 | {"cloud_topic": false, "delete_topic": true, "tick_interval": 3600000}            | integration | https://buildkite.com/redpanda/redpanda/builds/75946#019a6f71-0ce9-44a7-a007-82cbcacad3e5 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_working_node   |
| RedpandaNodeOperationsSmokeTest                   | test_node_ops_smoke_test                          | {"cloud_storage_type": 1, "mixed_versions": false}                                | integration | https://buildkite.com/redpanda/redpanda/builds/75946#019a6fca-66c3-4f29-84c5-a8a78a3f98b2 | FLAKY         | 20/21    | upstream reliability is '98.08743169398907'. current run reliability is '95.23809523809523'. drift is 2.84934 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test     |
| src/v/storage/mvlog/tests/file_test               | src/v/storage/mvlog/tests/file_test               |                                                                                   | unit        | https://buildkite.com/redpanda/redpanda/builds/75946#019a6f53-775d-45b4-9b18-690046cff551 | FAIL          | 0/1      |                                                                                                                                                                         |                                                                                                                                                                            |
| src/v/storage/mvlog/tests/file_test               | src/v/storage/mvlog/tests/file_test               |                                                                                   | unit        | https://buildkite.com/redpanda/redpanda/builds/75946#019a6f72-a180-44da-843c-26e6aa243555 | FAIL          | 0/1      |                                                                                                                                                                         |                                                                                                                                                                            |
| src/v/storage/mvlog/tests/file_test               | src/v/storage/mvlog/tests/file_test               |                                                                                   | unit        | https://buildkite.com/redpanda/redpanda/builds/75946#019a6f8e-c565-40d4-8626-d43e343c2d87 | FAIL          | 0/1      |                                                                                                                                                                         |                                                                                                                                                                            |
| src/v/storage/tests/segment_appender_rpbench_test | src/v/storage/tests/segment_appender_rpbench_test |                                                                                   | unit        | https://buildkite.com/redpanda/redpanda/builds/75946#019a6f53-775d-45b4-9b18-690046cff551 | FAIL          | 0/1      |                                                                                                                                                                         |                                                                                                                                                                            |
</details>
<!--/body-->

Lazin commented on src/v/storage/segment_appender.cc:
```diff
@@ -653,9 +653,11 @@ ss::future<> segment_appender::flush() {
-    return _out.flush().handle_exception([this](std::exception_ptr e) {
+    try {
+        return _out.flush().then([this] { ++_stats.fsyncs; });
+    } catch (const std::exception& e) {
```
the future is not co-awaited so the exception will not be handled by this try-catch block

WillemKauf replied:
`flush()` throws instead of returning an exceptional future I think, which would be caught by the `try/catch` block but not by a `handle_exception()` continuation

Lazin replied:
It'd be more readable with `co_await` and `try-catch`. Currently it relies on the assumption that internal implementation of the `ss::file` throws all exceptions before the first scheduling point. Without knowing about this assumption the code is difficult to understand. And the assumption couldn't be correct. I think this method eventually submits a syscall to the pool and returns a future which becomes ready when the syscall is executed. The only way for the `flush` operation to trigger an error is to do some I/O and hit EIO or something. It should happen after the scheduling point.

WillemKauf replied:
Right, the function itself is not a co-routine to begin with and I figure anything in a `segment_appender` should be considered a hot path and for that reason should _not_ be co-routinized, but I'm happy to make it a co-routine to deal with this issue if you'd prefer that.

WillemKauf replied:
Pushed

Lazin replied:
I mean both try-catch + co_await or continuation + handle_exception are fine.

WillemKauf replied:
This would have to be a `_out.flush().then(/*increment fsyncs*/).handle_exception();` block in the continuation version, which doesn't behave the way one would hope in handling exceptions.

WillemKauf committed:
`storage`: co-routinize `segment_appender::flush()`

WillemKauf committed:
`storage`: increment `fsyncs` in `flujsh()`

We increment `_stats.fsyncs` elsewhere we call `_out.flush()`, but not
at this call site.

Amend the call site to match the others.

This seems to have manifested in a test failure:

```
src/v/storage/tests/segment_appender_test.cc:237: Failure
Expected: (appender->get_stats().fsyncs) >= (1), actual: 0 vs 1
```

WillemKauf committed:
`storage`: remove check in `segment_appender_test`

Remove potentially flakey check on number of fsyncs being <= 2.
This occasionally flakes in CI/locally (about 1/1000 on my machine).

Code comment mentions the reason for this flake (occasionally redundant
flushes in `segment_appender`).

dotnwat commented on src/v/storage/segment_appender.cc:
```diff
@@ -630,11 +630,11 @@ ss::future<> segment_appender::flush() {
         // dispatching background head write.
         auto f = w.p.get_future();
         dispatch_background_head_write();
-        return f;
+        co_return co_await std::move(f);
```
it's probably fine, but did you consider any changes to scheduling points that might arise from a conversion to coroutines? since these code paths and the call sites of this predate coroutines by a long time the concern would be caller thinking something is running uninterrupted.

WillemKauf replied:
This PR got reverted back to the continuation form, just fyi: https://github.com/redpanda-data/redpanda/pull/28647

dotnwat replied:
are you sure that's the right PR? i don't see any un-coroutinization in there

WillemKauf replied:
Cover letter mentions the revert PR: https://github.com/redpanda-data/redpanda/pull/28486

dotnwat replied:
ahh ok i missed that. thanks. interesting!
